File: todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/tasks/TaskItemViewModelTest.java
/*
 * Copyright 2016, The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.android.architecture.blueprints.todoapp.tasks;

import android.content.Context;
import android.content.res.Resources;

import com.example.android.architecture.blueprints.todoapp.R;
import com.example.android.architecture.blueprints.todoapp.data.Task;
import com.example.android.architecture.blueprints.todoapp.data.source.TasksDataSource;
import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepository;

import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Unit tests for the implementation of {@link TasksViewModel}
 */
public class TaskItemViewModelTest {

    private static final String NO_DATA_STRING = "NO_DATA_STRING";

    private static final String NO_DATA_DESC_STRING = "NO_DATA_DESC_STRING";

    @Mock
    private TasksRepository mTasksRepository;

    @Mock
    private Context mContext;

    @Mock
    private TaskItemNavigator mTaskItemNavigator;

    @Captor
    private ArgumentCaptor<TasksDataSource.GetTaskCallback> mLoadTasksCallbackCaptor;

    private TaskItemViewModel mTaskItemViewModel;

    private Task mTask;

    @Before
    public void setupTasksViewModel() {
        // Mockito has a very convenient way to inject mocks by using the @Mock annotation. To
        // inject the mocks in the test the initMocks method needs to be called.
        MockitoAnnotations.initMocks(this);

        setupContext();

        // Get a reference to the class under test
        mTaskItemViewModel = new TaskItemViewModel(mContext, mTasksRepository, mTaskItemNavigator);

    }

    private void setupContext() {
        when(mContext.getApplicationContext()).thenReturn(mContext);
        when(mContext.getString(R.string.no_data)).thenReturn(NO_DATA_STRING);
        when(mContext.getString(R.string.no_data_description)).thenReturn(NO_DATA_DESC_STRING);

        when(mContext.getResources()).thenReturn(mock(Resources.class));
    }

    @Test
    public void clickOnTask_ShowsDetailUi() {
        loadTaskIntoViewModel();

        mLoadTasksCallbackCaptor.getValue().onTaskLoaded(mTask); // Trigger callback

        // Then task detail UI is shown
        assertEquals(mTaskItemViewModel.getTitle(), mTask.getTitle());
        assertEquals(mTaskItemViewModel.getDescription(), mTask.getDescription());
    }

    @Test
    public void nullTask_showsNoData() {
        loadTaskIntoViewModel();

        mLoadTasksCallbackCaptor.getValue().onTaskLoaded(null); // Trigger callback

        // Then task detail UI is shown
        assertEquals(mTaskItemViewModel.getTitle(), NO_DATA_STRING);
        assertEquals(mTaskItemViewModel.getDescription(), NO_DATA_DESC_STRING);
    }

    @Test
    public void completeTask_ShowsTaskMarkedComplete() {
        loadTaskIntoViewModel();

        mLoadTasksCallbackCaptor.getValue().onTaskLoaded(mTask); // Trigger callback

        // When task is marked as complete
        mTaskItemViewModel.setCompleted(true);

        // Then repository is called
        verify(mTasksRepository).completeTask(mTask);
    }

    @Test
    public void activateTask_ShowsTaskMarkedActive() {
        loadTaskIntoViewModel();

        mLoadTasksCallbackCaptor.getValue().onTaskLoaded(mTask); // Trigger callback

        // When task is marked as complete
        mTaskItemViewModel.setCompleted(false);

        // Then repository is called
        verify(mTasksRepository).activateTask(mTask);
    }

    @Test
    public void unavailableTasks_ShowsError() {
        loadTaskIntoViewModel();

        mLoadTasksCallbackCaptor.getValue().onDataNotAvailable(); // Trigger callback

        // Then repository is called
        assertFalse(mTaskItemViewModel.isDataAvailable());
    }

    private void loadTaskIntoViewModel() {
        // Given a stubbed active task
        mTask = new Task("Details Requested", "For this task");

        // When open task details is requested
        mTaskItemViewModel.start(mTask.getId());

        // Use a captor to get a reference for the callback.
        verify(mTasksRepository).getTask(eq(mTask.getId()), mLoadTasksCallbackCaptor.capture());
    }
}

File: todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/tasks/TaskItemViewModelTest.kt
/*
 * Copyright 2017, The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.android.architecture.blueprints.todoapp.tasks

import android.content.Context
import android.content.res.Resources
import com.example.android.architecture.blueprints.todoapp.R
import com.example.android.architecture.blueprints.todoapp.data.Task
import com.example.android.architecture.blueprints.todoapp.data.source.TasksDataSource
import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepository
import com.example.android.architecture.blueprints.todoapp.util.capture
import com.example.android.architecture.blueprints.todoapp.util.eq
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Before
import org.junit.Test
import org.mockito.ArgumentCaptor
import org.mockito.Captor
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations

/**
 * Unit tests for the implementation of [TasksViewModel]
 */
class TaskItemViewModelTest {

    @Mock private lateinit var tasksRepository: TasksRepository
    @Mock private lateinit var context: Context
    @Mock private lateinit var taskItemNavigator: TasksActivity
    @Captor private lateinit var loadTasksCallbackCaptor: ArgumentCaptor<TasksDataSource.GetTaskCallback>
    private lateinit var taskItemViewModel: TaskItemViewModel
    private lateinit var task: Task
    private val NO_DATA_STRING = "NO_DATA_STRING"
    private val NO_DATA_DESC_STRING = "NO_DATA_DESC_STRING"

    @Before fun setupTasksViewModel() {
        // Mockito has a very convenient way to inject mocks by using the @Mock annotation. To
        // inject the mocks in the test the initMocks method needs to be called.
        MockitoAnnotations.initMocks(this)

        setupContext()

        // Get a reference to the class under test
        taskItemViewModel = TaskItemViewModel(context, tasksRepository).apply {
            setNavigator(taskItemNavigator)
        }
    }

    private fun setupContext() {
        `when`(context.applicationContext).thenReturn(context)
        `when`(context.getString(R.string.no_data)).thenReturn(NO_DATA_STRING)
        `when`(context.getString(R.string.no_data_description)).thenReturn(NO_DATA_DESC_STRING)
        `when`(context.resources).thenReturn(mock(Resources::class.java))
    }

    @Test fun clickOnTask_ShowsDetailUi() {
        loadTaskIntoViewModel()

        loadTasksCallbackCaptor.value.onTaskLoaded(task) // Trigger callback

        // Then task detail UI is shown
        assertEquals(taskItemViewModel.title.get(), task.title)
        assertEquals(taskItemViewModel.description.get(), task.description)
    }

    @Test fun completeTask_ShowsTaskMarkedComplete() {
        loadTaskIntoViewModel()

        loadTasksCallbackCaptor.value.onTaskLoaded(task) // Trigger callback

        // When task is marked as complete
        taskItemViewModel.completed = true

        // Then repository is called
        verify<TasksRepository>(tasksRepository).completeTask(task)
    }

    @Test fun activateTask_ShowsTaskMarkedActive() {
        loadTaskIntoViewModel()

        loadTasksCallbackCaptor.value.onTaskLoaded(task) // Trigger callback

        // When task is marked as complete
        taskItemViewModel.completed = false

        // Then repository is called
        verify<TasksRepository>(tasksRepository).activateTask(task)
    }

    @Test fun unavailableTasks_ShowsError() {
        loadTaskIntoViewModel()

        loadTasksCallbackCaptor.value.onDataNotAvailable() // Trigger callback

        // Then repository is called
        assertFalse(taskItemViewModel.isDataAvailable)
    }

    private fun loadTaskIntoViewModel() {
        // Given a stubbed active task
        task = Task("Details Requested", "For this task")

        // When open task details is requested
        taskItemViewModel.start(task.id)

        // Use a captor to get a reference for the callback.
        verify<TasksRepository>(tasksRepository).getTask(eq(task.id),
                capture(loadTasksCallbackCaptor))
    }
}
